### PR TITLE
Add end locations to statements

### PR DESCRIPTION
**Pull request by sim642**
This is a non-heuristic implementation of https://github.com/goblint/analyzer/issues/428.

It was easiest to add corresponding end fields to the Cabs and Cil location records instead of changing the location types entirely to be pairs of locations (start, end). This has minimal compatibility problems.

Single locations (with unknown end) are joined together into ranges for parsed statements, but nothing else. In Goblint we do everything, including location tracking, per-statement anyway, so this should be good enough for now.
If we wanted more precise locations, we'd have to change Cil to have location (ranges) attached to all subexpressions, which would be a massive breaking change.

### CODE REVIEW

**Comment by sim642** [comment]
Initially the location range for a non-primitive statement (e.g. `if`, `while`, etc) was the entire statement, but this turned out to be very inconvenient in Goblint, because warnings about the condition spanned the entire body of the statement as well.
I now added additional just expression locations to those statements where it's useful to be able to just get the location range for the expression.

**EDIT:** Also had to add secondary locations to primitive statements because some of them may originate from temporary assignments and calls CIL inserts before a conditional. Those should still place the warning on the corresponding conditional expression they're part of.

File: src/frontc/cabs2cil.ml [hunk 1]
```diff
@@ -1181,6 +1174,7 @@ class registerLabelsVisitor = object
 
   method! vstmt s =
     currentLoc := convLoc (C.get_statementloc s);
+    (* don't touch currentExpLoc! *)
```
Comment (michael-schwarz) [inline src/frontc/cabs2cil.ml]: Can you elaborate on why one should not touch it here?

File: src/frontc/cabs2cil.ml [hunk 2]
```diff
@@ -5804,6 +5798,7 @@ and doAliasFun vtype (thisname:string) (othername:string)
 and doDecl (isglobal: bool) : A.definition -> chunk = function
   | A.DECDEF ((s, nl), loc) ->
       currentLoc := convLoc(loc);
+      currentExpLoc := convLoc loc; (* eloc for local initializer assignment instruction *)
```
Comment (michael-schwarz) [inline src/frontc/cabs2cil.ml]: Having the same thing right below each other one time in `(...)` and one time without it is a bit ugly.
I'd suggest to remove the parens in the line above.

**Review by michael-schwarz — APPROVED** [review]

File: src/frontc/cabs2cil.ml (hunk 1 above)
Comment (sim642) [reply src/frontc/cabs2cil.ml]: I just extended the comments related to this. There's no deep reason other than if you change it here, it also needs to be reset after using the visitor in `doDecl`. The comment is mainly here as a reminder not to touch it here without resetting it afterwards and possibly breaking something.

**Comment by michael-schwarz** [comment]
Did you investigate whether this breaks Gobview before merging?

**Comment by michael-schwarz** [comment]
Did you investigate whether this breaks Gobview before merging?

**Comment by sim642** [comment]
> Did you investigate whether this breaks Gobview before merging?

Hmm, no, but tried it right now and it compiled without errors. Adding the end fields doesn't really break things, but Gobview simply doesn't use them. Given https://github.com/goblint/gobview/issues/2, there's probably no use for them anyway right now.

**Pull request merged**